cylc/flow/task_outputs.py

```diff
@@ -44,15 +67,207 @@
     TASK_OUTPUT_FINISHED,
 )
 
-_TRIGGER = 0
-_MESSAGE = 1
-_IS_COMPLETED = 2
+# this evaluates task completion expressions
+CompletionEvaluator = restricted_evaluator(
+    # expressions
+    ast.Expression,
+    # variables
+    ast.Name, ast.Load,
+    # operations
+    ast.BoolOp, ast.And, ast.Or, ast.BinOp,
+    error_class=InvalidCompletionExpression,
+)
+
+# regex for splitting expressions into individual parts for formatting
+RE_EXPR_SPLIT = re.compile(r'([\(\) ])')
+
+
+def trigger_to_completion_variable(output: str) -> str:
+    """Turn a trigger into something that can be used in an expression.
+
+    Examples:
+        >>> trigger_to_completion_variable('succeeded')
+        'succeeded'
+        >>> trigger_to_completion_variable('submit-failed')
+        'submit_failed'
+
+    """
+    return output.replace('-', '_')
+
+
+def completion_variable_to_trigger(compvar: str) -> str:
+    """Turn a completion variable into a trigger.
+
+    Examples:
+        >>> completion_variable_to_trigger('succeeded')
+        'succeeded'
+        >>> completion_variable_to_trigger('submit_failed')
+        'submit-failed'
+
+    """
+    return compvar.replace('_', '-')
```

Comment:
Then we can't reverse the process :(

This function is only used to format user exceptions for validate-time config issues. So it's non functional, but it would be good to give the error back to the user in the format they gave it in. If we can't easily get the trigger from the completion variable, then I'll strip this.